### PR TITLE
[CW-1625] Added DeleteFilter to Tx interfaces

### PR DESCRIPTION
**Pull request by zimnx**
- Added StateList to Tx interface
- ResourceState contains information about ID of resource it points to
- Added unit tests to transaction_event_logger
- fixed failing test on SQLite

### CODE REVIEW

**Comment by zimnx** [comment]
Tests locally passed. There might be issue with MariaDB which is returning float64 for int64 types. I'm still investigating it, but please don't hesitate to review it :)

File: tests/test_schema.yaml [hunk 1]
```diff
@@ -1655,7 +1655,7 @@ schemas:
   prefix: /v2.0
   schema:
     indexes:
-      test_unique_id_and_tenant_id_and_domain_id:
+      custom_action_test_unique_id_and_tenant_id_and_domain_id:
```
Comment (jkmar) [inline tests/test_schema.yaml]: why is this changed?

File: server/transaction_event_logger_test.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	)
+
+	withinTx := func(fn func(transaction.Transaction)) {
+		Expect(db.WithinTx(syncedDb, func(tx transaction.Transaction) error {
+			fn(tx)
+			return nil
+		})).To(Succeed())
+	}
+
+	deleteAllEvents := func() {
```
Comment (jkmar) [inline server/transaction_event_logger_test.go]:
```go
deleteAll := func(table string) {
    withinTx(func(tx transaction.Transaction) {
        Expect(tx.Exec(ctx, "DELETE FROM " + table)).To(Succeed())
    })
}

deleteAllNetworks := func() { deleteAll("networks") }
deleteAllEvents := func() { deleteAll("events") }
```

File: server/transaction_event_logger_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	deleteAllNetworks := func() {
+		withinTx(func(tx transaction.Transaction) {
+			Expect(tx.Exec(ctx, "DELETE FROM networks")).To(Succeed())
+		})
+	}
+
+	BeforeEach(func() {
+		// go vet complains about cancel(), but it's called in AfterEach
```
Comment (jkmar) [inline server/transaction_event_logger_test.go]: Does this comment still apply?

File: server/server_test.go [hunk 4]
```diff
@@ -1922,6 +1922,13 @@ func getNetwork(color string, tenant string) map[string]interface{} {
 	}
 }
 
+func getSchema(schemaID string) *schema.Schema {
```
Comment (jkmar) [inline server/server_test.go]: Why isn't this function defined in the same file it's used?

File: server/transaction_event_logger_test.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		ctx, cancel = context.WithCancel(context.Background())
+
+		syncedDb = srv.NewDbSyncWrapper(testDB)
+		eventSchema, _ = schema.GetManager().Schema("event")
```
Comment (jkmar) [inline server/transaction_event_logger_test.go]: `getSchema`?

File: server/transaction_event_logger_test.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+			createNetwork(ctx, tx, "red")
+			createNetwork(ctx, tx, "green")
+			createNetwork(ctx, tx, "blue")
+
+			Expect(listEvents(tx)).To(HaveLen(3))
+
+			deleteAllFilter := filter.True()
+			Expect(tx.DeleteFilter(ctx, getSchema("network"), deleteAllFilter)).To(Succeed())
+
+			Expect(listEvents(tx)).To(HaveLen(6))
```
Comment (jkmar) [inline server/transaction_event_logger_test.go]: Maybe also check event types like in the previous test?

**Comment by zimnx** [comment]
> Tests locally passed. There might be issue with MariaDB which is returning float64 for int64 types. I'm still investigating it, but please don't hesitate to review this change ;)

Already fixed. DB driver was returning []uint8 instead int64. For fix see newest commit.

File: db/sql/filter.go [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+)
+
+const (
+	OrCondition   = "__or__"
+	AndCondition  = "__and__"
+	BoolCondition = "__bool__"
+)
+
+type queryBuilder interface {
+	Where(pred interface{}, args ...interface{})
```
Comment (jkmar) [inline db/sql/filter.go]: Why not `Where (pred interface{}, args ...interface{}) queryBuilder`, to be consistent with the squirrel library?

File: tests/test_schema.yaml (hunk 1 above)
Comment (zimnx) [reply tests/test_schema.yaml]: to fix failing tests on SQLite. Current master is broken.

File: server/transaction_event_logger_test.go (hunk 3 above)
Comment (zimnx) [reply server/transaction_event_logger_test.go]: yes

File: server/transaction_event_logger_test.go (hunk 2 above)
Comment (zimnx) [reply server/transaction_event_logger_test.go]: done

File: server/server_test.go (hunk 4 above)
Comment (zimnx) [reply server/server_test.go]: moved

File: server/transaction_event_logger_test.go (hunk 5 above)
Comment (zimnx) [reply server/transaction_event_logger_test.go]: done

File: server/transaction_event_logger_test.go (hunk 6 above)
Comment (zimnx) [reply server/transaction_event_logger_test.go]: done

File: db/sql/filter.go (hunk 7 above)
Comment (zimnx) [reply db/sql/filter.go]: I recall I started with this approach but for some reason I chose to remove return value. I'm not going to discover it again.

File: db/sql/sql_test.go [hunk 8]
```diff
@@ -766,14 +766,16 @@ var _ = Describe("Sql", func() {
...
-					res, err := AddFilterToQuery(testSchema, query, filter, false)
+					res, err := AddFilterToSelectQuery(testSchema, query, filter, false)
 
 					Expect(err).ToNot(HaveOccurred())
 					resSql, param, err := res.ToSql()
 					Expect(err).ToNot(HaveOccurred())
 
 					expectedQuery = expectedQuery.Where(expected)
 					expectedSql, expectedParam, _ := expectedQuery.ToSql()
+					fmt.Println("XD", resSql, expectedSql)
```
Comment (tomasz-michalski) [inline db/sql/sql_test.go]: I think this should be removed

File: db/transaction/fuzzy_transaction.go [hunk 9]
```diff
@@ -90,6 +95,16 @@ func (ft *FuzzyTransaction) StateFetch(ctx context.Context, s *schema.Schema, fi
 	})
 }
 
+// List lists resources
```
Comment (tomasz-michalski) [inline db/transaction/fuzzy_transaction.go]: This comment should be updated

File: extension/goplugin/schemas.go [hunk 10]
```diff
@@ -458,25 +464,38 @@ func (schema *Schema) DbStateUpdateRaw(rawResource interface{}, requestContext g
...
+	return schema.dbDeleteFilter(goext.Filter{"id": id}, context)
 }
 
 // DbDeleteFilterRaw deletes resource by filter without triggering events
 func (schema *Schema) DbDeleteFilterRaw(filter goext.Filter, context goext.Context) error {
-	return schema.delete(filter, context, false)
+	return schema.dbDeleteFilter(filter, context)
+}
+
+func (schema *Schema) dbDeleteFilter(filter goext.Filter, ctx goext.Context) error {
```
Comment (tomasz-michalski) [inline extension/goplugin/schemas.go]: Please rename `dbDeleteFilter` to `dbDelete` or rename `delete` to `deleteFilter`, because now these names are inconsistent.

File: db/db_test.go [hunk 11]
```diff
@@ -391,6 +391,12 @@ var _ = Describe("Database operation test", func() {
 					Expect(tx.Commit()).To(Succeed())
 				})
 
+				It("Deletes resources using filter", func() {
+					Expect(tx.DeleteFilter(ctx, serverSchema, transaction.IDFilter(serverResource.ID()))).To(Succeed())
```
Comment (michalzurawski) [inline db/db_test.go]: this test will pass with empty body of DeleteFilter method i.e. it doesn't check if resource from filter were deleted

File: server/transaction_event_logger.go [hunk 12]
```diff
@@ -149,6 +153,50 @@ func (tl *transactionEventLogger) Delete(ctx context.Context, s *schema.Schema,
 	return tl.logEvent(ctx, "delete", resource, configVersion)
 }
 
+func (tl *transactionEventLogger) DeleteFilter(ctx context.Context, s *schema.Schema, filter transaction.Filter) error {
+	if !tl.syncEventNeeded(s) {
```
Comment (tomasz-michalski) [inline server/transaction_event_logger.go]: This is just a fast path, without it result would be the same (although slower). Am I right?

File: db/sql/filter.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+			}
+
+			var column string
+			if join {
+				column = makeColumn(s.GetDbTableName(), *property)
+			} else {
+				column = quote(key)
+			}
+
+			// TODO: add other operators
```
Comment (michalzurawski) [inline db/sql/filter.go]: nit: this TODO comment was referring to add more types like "eq" and "neq" in line 165 - please either move it there or either remove at all as now it's useless. And yes, i'm aware that you just moved this function to separate file (which is good idea btw)

File: server/transaction_event_logger.go (hunk 12 above)
Comment (zimnx) [reply server/transaction_event_logger.go]: Indeed

File: extension/goplugin/schemas.go (hunk 10 above)
Comment (zimnx) [reply extension/goplugin/schemas.go]: done

File: db/transaction/fuzzy_transaction.go (hunk 9 above)
Comment (zimnx) [reply db/transaction/fuzzy_transaction.go]: done

File: db/sql/sql_test.go (hunk 8 above)
Comment (zimnx) [reply db/sql/sql_test.go]: done

File: db/sql/filter.go (hunk 13 above)
Comment (zimnx) [reply db/sql/filter.go]: done

File: db/db_test.go (hunk 11 above)
Comment (zimnx) [reply db/db_test.go]: done

File: db/transaction/fuzzy_transaction.go [hunk 14]
```diff
@@ -60,6 +60,11 @@ func (ft *FuzzyTransaction) Delete(ctx context.Context, s *schema.Schema, resour
 	return ft.fuzzIt(func() error { return ft.Tx.Delete(ctx, s, resourceID) })
 }
 
+// Delete deletes a resource
```
Comment (tomasz-michalski) [inline db/transaction/fuzzy_transaction.go]: `// DeleteFilter deletes a resource`

File: db/transaction/fuzzy_transaction.go [hunk 15]
```diff
@@ -90,6 +95,16 @@ func (ft *FuzzyTransaction) StateFetch(ctx context.Context, s *schema.Schema, fi
 	})
 }
 
+// List lists resource states
```
Comment (tomasz-michalski) [inline db/transaction/fuzzy_transaction.go]: `// StateList lists resource states`
